torch/_functorch/vmap.py

```diff
@@ -391,19 +398,114 @@ def vmap(
 
     @functools.wraps(func)
     def wrapped(*args, **kwargs):
         lazy_load_decompositions()
         _check_out_dims_is_int_or_int_pytree(out_dims, func)
         batch_size, flat_in_dims, flat_args, args_spec = _process_batched_inputs(in_dims, args, func)
+
+        if chunk_size is not None:
+            chunks_flat_args = _get_chunked_inputs(flat_args, flat_in_dims, batch_size, chunk_size)
+            return _chunked_vmap(func, flat_in_dims, chunks_flat_args,
+                                 args_spec, out_dims, randomness, **kwargs)
+
+        # If chunk_size is not specified.
         return _flat_vmap(
             func, batch_size, flat_in_dims, flat_args, args_spec, out_dims, randomness, **kwargs
         )
 
     return wrapped
 
 
+def _flatten_chunks_output(chunks_output_):
+    # chunks_output is a list of chunked outputs
+    # flatten chunked outputs:
+    flat_chunks_output = []
+    arg_spec = None
+    for output in chunks_output_:
+        flat_output, arg_specs = tree_flatten(output)
+        flat_chunks_output.append(flat_output)
+        if arg_spec is None:
+            arg_spec = arg_specs
+
+    # transpose chunk dim and flatten structure
+    # flat_output_chunks is flat list of chunks
+    flat_output_chunks = list(zip(*flat_chunks_output))
+    return flat_output_chunks, arg_spec
+
+
+def _concat_chunked_outputs(out_dims, arg_spec, flat_output_chunks):
+    # concat chunks on out_dim
+    flat_out_dims = _broadcast_to_and_flatten(out_dims, arg_spec)
+    assert len(flat_out_dims) == len(flat_output_chunks)
+    flat_output = []
+    for out_dim in flat_out_dims:
+        # Here, idx of `flat_output_chunks` is always `0`.
+        # This is because we delete the `0` index which shifts
+        # all the remaining elements to the left.
+        # Eg.
+        # >>> a = [1, 2, 3]
+        # >>> del a[0]  # a = [2, 3]
+        # >>> del a[0]  # a = [3]
+        flat_output.append(torch.cat(flat_output_chunks[0], dim=out_dim))
+        # release source data
+        del flat_output_chunks[0]
+    del flat_output_chunks
+
+    return flat_output
+
+
+# Applies vmap on chunked_input and returns concatenated output over the chunks.
+def _chunked_vmap(func, flat_in_dims, chunks_flat_args, args_spec, out_dims, randomness, **kwargs):
+    chunks_output = []
+    rs = torch.get_rng_state() if randomness == "same" else None
+    for flat_args in chunks_flat_args:
+        batch_size = _validate_and_get_batch_size(flat_in_dims, flat_args)
+        if batch_size == 0:
+            continue
+
+        if rs is not None:
+            torch.set_rng_state(rs)
+        chunks_output.append(
+            _flat_vmap(
+                func, batch_size, flat_in_dims, flat_args, args_spec, out_dims, randomness, **kwargs
+            )
+        )
+
+    flat_output_chunks, arg_spec = _flatten_chunks_output(chunks_output)
+
+    # Removing temporary variables helps to reduce memory usage on device like CUDA
+    del chunks_output
```

Comment:
To check my understanding:
- we del the Tensors inside `_concat_chunked_outputs` and we need `del chunks_output` because that holds a reference to those tensors?

If so, we should put that into the comment that the Tensors aren't immediately freed at `del chunks_output` -- that was what confused me

Reply:
Updated.